pkg/controller/volume/persistentvolume/scheduler_binder_test.go

```diff
@@ -1383,6 +1402,38 @@ func TestBindPodVolumes(t *testing.T) {
 			},
 			shouldFail: true,
 		},
+		"node-affinity-check-fails": {
+			binding:   binding1aBound,
+			cachedPV:  pvNode1a,
+			cachedPVC: unboundPVC,
+			delayFunc: func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim) {
+				// Update PVC to be fully bound to PV
+				newPVC := pvc.DeepCopy()
+				newPVC.ResourceVersion = "100"
+				newPVC.Spec.VolumeName = pv.Name
+				metav1.SetMetaDataAnnotation(&newPVC.ObjectMeta, annBindCompleted, "yes")
+
+				// Update pvc cache, fake client doesn't invoke informers
+				internalBinder, ok := testEnv.binder.(*volumeBinder)
+				if !ok {
+					t.Fatalf("Failed to convert to internal binder")
+				}
+
+				pvcCache := internalBinder.pvcCache
+				internalPVCCache, ok := pvcCache.(*pvcAssumeCache)
+				if !ok {
+					t.Fatalf("Failed to convert to internal PVC cache")
+				}
+				internalPVCCache.add(newPVC)
+
+				// Update pod binding cache
```

Comment:
why does pod binding cache need to be updated?  `checkBoundClaims` should only need to look at the pvc and pv cache

Reply:
Ah I see, in that case, it should be added to pod binding cache during test setup, not the delay func.  The testEnv.assumeVolumes call currently passes in no PVCs, but it can be modified.

Reply:
It seems like `testEnv.assumeVolumes` do passes in PVCs. I missed that.

Unfortunately I can't update the pod binding cache in all cases, so I see a few options to solve this. I implemented the third one, but let me know if you have a better idea:

1. Update the pod binding cache only in the delay function (which we don't want)
2. Add a flag in the test case to update the pod binding cache (which we probably don't want as well, as per a previous comment in this PR)
3. Add to the test case a list of provisioned PVCs, just like other tests are doing :)

Just for the record, here's why I can't update the pod binding cache in all cases: `boundPVCNode1a` doesn't have the field `boundPVCNode1a.Spec.VolumeName` set, and it seems like that's intentional [1]. However, if I update the pod binding cache without this field set, this [2] validation in `volumeBinder.checkBindings` will fail. I believe that's the main reason why this test isn't updating the cache.

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/scheduler_binder_test.go#L452-L461
[2] https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/scheduler_binder.go#L418-L421

Reply:
Can we do something like this?
* the test PVC has selectedNode set to nodeX, not bound
* delayFunc updates the PVC to be bound to the wrong PV with different node

Reply:
Do you mean something like this?

Reply:
Done